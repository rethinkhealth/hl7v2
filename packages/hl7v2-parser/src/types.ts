export interface ParseOptions {
  delimiters?: Partial<HL7v2Delimiters>;
  autoDetectDelimiters?: boolean;
}

/**
 * HL7v2 Delimiters type
 */
export interface HL7v2Delimiters {
  field: string;
  component: string;
  subcomponent: string;
  repetition: string;
  escape: string;
  segment: string; // now always a string
}
