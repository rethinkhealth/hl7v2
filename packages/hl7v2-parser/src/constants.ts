import type { HL7v2Node } from '@rethinkhealth/hl7v2-ast';
import type { HL7v2Delimiters } from './types';

/**
 * Default HL7v2 delimiters
 */
export const DEFAULT_DELIMITERS: HL7v2Delimiters = {
  field: '|',
  component: '^',
  subcomponent: '&',
  repetition: '~',
  escape: '\\',
  segment: '\r', // default to \r
};

/**
 * Empty message node
 */
export const EMPTY_MESSAGE: HL7v2Node = {
  type: 'root',
  delimiter: DEFAULT_DELIMITERS.segment,
  children: [],
  position: {
    start: { line: 1, column: 1, offset: 0 },
    end: { line: 1, column: 1, offset: 0 },
  },
};
