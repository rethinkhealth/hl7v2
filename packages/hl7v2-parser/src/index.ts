// src/index.ts

export { default as hl7v2Parser } from './parser';
export * from './preprocessor';
export { parseHL7v2FromIterator } from './processor';
export * from './tokenizer';
export * from './types';
