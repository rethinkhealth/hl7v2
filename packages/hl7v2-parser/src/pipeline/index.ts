// Core pipeline components
export { HL7MessageParser } from './core/message';
export { IComponentParser as DefaultComponentParser } from './core/parsers/component';
export { FieldParser as DefaultFieldParser } from './core/parsers/field';
// Built-in parsers
export { SegmentParser as DefaultSegmentParser } from './core/parsers/segment';
export { MSHSegmentParser } from './core/parsers/segment-msh';
export { SubcomponentParser as DefaultSubcomponentParser } from './core/parsers/subcomponent';
export { SegmentParserRegistry } from './core/registry';
// Interfaces
export type {
  ComponentParseInput,
  FieldParseInput,
  IComponentParser as ComponentParser,
  IFieldParser as FieldParser,
  IMessageParser as MessageParser,
  ISegmentParser as SegmentParser,
  ISubcomponentParser as SubcomponentParser,
  MessageParseInput,
  ParseContext,
  ParseInput,
  ParseOptions,
  Parser,
  PipelineStage,
  SegmentParseInput,
  SubcomponentParseInput,
  ValidationHook,
  ValidationResult,
} from './interfaces';
// Validation hooks
export {
  BasicValidationHook,
  SegmentRequirementValidationHook,
} from './validation/basic-validation';
