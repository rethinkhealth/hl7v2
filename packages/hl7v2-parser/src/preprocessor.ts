import { DEFAULT_DELIMITERS } from "@rethinkhealth/hl7v2-utils";
import type { ParserContext, PreprocessorStep } from "./types";

// PreprocessorStep is declared in `types.ts` to avoid circular imports

/**
 * Step: strip UTF-8 BOM.
 */
// biome-ignore lint/style/useNamingConvention: fine
export const stripBOM: PreprocessorStep = (ctx) => {
  if (ctx.input.charCodeAt(0) === 0xfe_ff) {
    ctx.input = ctx.input.slice(1);
  }
  return ctx;
};

/**
 * Step: normalize newlines to the default delimiters.
 */
export const normalizeNewlines: PreprocessorStep = (ctx) => {
  ctx.input = ctx.input.replace(/\r?\n/g, ctx.delimiters.segment);
  return ctx;
};

/**
 * Step: auto-detect delimiters from MSH.
 */
export const detectDelimiters: PreprocessorStep = (ctx) => {
  if (ctx.input.startsWith("MSH")) {
    const field = ctx.input.charAt(3) || DEFAULT_DELIMITERS.field;
    const enc = ctx.input.slice(4, 8);
    const component = enc.charAt(0) || DEFAULT_DELIMITERS.component;
    const repetition = enc.charAt(1) || DEFAULT_DELIMITERS.repetition;
    const _escape = enc.charAt(2) || DEFAULT_DELIMITERS.escape;
    const subcomponent = enc.charAt(3) || DEFAULT_DELIMITERS.subcomponent;
    const segment = DEFAULT_DELIMITERS.segment;

    ctx.delimiters = {
      field,
      component,
      repetition,
      escape: _escape,
      subcomponent,
      segment,
    };
  }
  return ctx;
};

/**
 * Default preprocessing pipeline.
 */
export const defaultPreprocessors: PreprocessorStep[] = [
  stripBOM,
  normalizeNewlines,
  detectDelimiters,
];

/**
 * Run the pipeline to initialize ParserContext.
 */
export function runPreprocessors(
  ctx: ParserContext,
  steps: PreprocessorStep[]
): ParserContext {
  for (const step of steps) {
    // biome-ignore lint/style/noParameterAssign: this is necessary to keep no-copy
    ctx = step(ctx);
  }

  return ctx;
}
