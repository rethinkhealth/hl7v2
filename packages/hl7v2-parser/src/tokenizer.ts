// src/tokenizer.ts
import type { HL7v2Delimiters } from '@rethinkhealth/hl7v2-utils';
import type {
  ParserContext,
  Position,
  Token,
  Tokenizer,
  TokenType,
} from './types';

const MSH_SEGMENT_START = 0;
const MSH_SEGMENT_END = 3;
const MSH_FIELD_SEPERATOR_START = 3;
const MSH_FIELD_SEPERATOR_END = 4;
const MSH_FIELD_DELIMITER_START = 4;
const MSH_FIELD_DELIMITER_END = 8;

export class HL7v2Tokenizer implements Tokenizer, Iterable<Token> {
  private input = '';
  private i = 0;
  private line = 1;
  private col = 1;
  private delims!: HL7v2Delimiters;

  // Only-run-once MSH bootstrap at the start of the file
  private didMshBootstrap = false;
  private pendingBootstrap: null | Array<
    | { kind: 'TEXT'; value: string; advance: number }
    | { kind: 'FIELD_DELIM'; advance: number }
  > = null; // queue to emit TEXT('MSH'), FIELD_DELIM, TEXT('^~\\&')

  reset(ctx: ParserContext) {
    this.input = ctx.input;
    this.delims = ctx.delimiters;
    this.i = 0;
    this.line = 1;
    this.col = 1;
    this.didMshBootstrap = false;
    this.pendingBootstrap = null;

    // Prepare a one-time bootstrap if file starts with MSH
    if (this.input.startsWith('MSH')) {
      // Precompute MSH and MSH.2; MSH.1 is the field delimiter char at index 3
      const msh = this._slice(MSH_SEGMENT_START, MSH_SEGMENT_END);
      const msh1 = this._slice(
        MSH_FIELD_SEPERATOR_START,
        MSH_FIELD_SEPERATOR_END
      ); // the field delimiter char at index 3
      const msh2 = this._slice(
        MSH_FIELD_DELIMITER_START,
        MSH_FIELD_DELIMITER_END
      ); // may be shorter than 4 if truncated
      this.pendingBootstrap = [
        { kind: 'TEXT', value: msh, advance: msh.length },
        { kind: 'FIELD_DELIM', advance: 0 },
        { kind: 'TEXT', value: msh1, advance: msh1.length },
        { kind: 'FIELD_DELIM', advance: 0 },
        { kind: 'TEXT', value: msh2, advance: msh2.length },
      ];
      // NOTE: we have not advanced indices yet; we will as we emit tokens
    }
  }

  // biome-ignore lint/complexity/noExcessiveCognitiveComplexity: The cognitive complexity of this method is justified because it must handle the HL7v2 MSH segment bootstrap logic and multiple tokenization cases in a single method for performance and maintainability.
  next(): Token | null {
    const s = this.input;
    const n = s.length;
    if (this.i >= n && !this.pendingBootstrap?.length) {
      return null;
    }

    const start = { offset: this.i, line: this.line, column: this.col };

    // ---- One-time MSH bootstrap at file start ----
    if (!this.didMshBootstrap && this.pendingBootstrap) {
      const step = this.pendingBootstrap.shift();
      if (step) {
        if (step.kind === 'TEXT') {
          const take = Math.min(step.advance, n - this.i);
          const out = step.value.slice(0, take);
          this._fastAdvance(out);
          if (this.pendingBootstrap.length === 0) {
            this.pendingBootstrap = null;
            this.didMshBootstrap = true;
          }
          return this._tok('TEXT', out, start);
        }
        // FIELD_DELIM: advance exactly one char (the delimiter) and emit a FIELD_DELIM token
        this._advance(Math.min(step.advance, n - this.i));
        if (this.pendingBootstrap.length === 0) {
          this.pendingBootstrap = null;
          this.didMshBootstrap = true;
        }
        return this._tok('FIELD_DELIM', undefined, start);
      }
    }

    // ---- Normal tokenization (delimiters + text) ----
    if (this.i >= n) {
      return null;
    }
    const ch = s.charAt(this.i);

    if (ch === this.delims.segment) {
      this._advance(1, true);
      return this._tok('SEGMENT_END', undefined, start);
    }
    if (ch === this.delims.field) {
      this._advance(1);
      return this._tok('FIELD_DELIM', undefined, start);
    }
    if (ch === this.delims.repetition) {
      this._advance(1);
      return this._tok('REPETITION_DELIM', undefined, start);
    }
    if (ch === this.delims.component) {
      this._advance(1);
      return this._tok('COMPONENT_DELIM', undefined, start);
    }
    if (ch === this.delims.subcomponent) {
      this._advance(1);
      return this._tok('SUBCOMP_DELIM', undefined, start);
    }

    // TEXT until next delimiter or end
    let j = this.i;
    const seg = this.delims.segment,
      fld = this.delims.field,
      rep = this.delims.repetition,
      cmp = this.delims.component,
      sub = this.delims.subcomponent;

    while (j < s.length) {
      const c = s.charAt(j);
      if (c === seg || c === fld || c === rep || c === cmp || c === sub) {
        break;
      }
      j++;
    }

    const val = s.slice(this.i, j);
    this._fastAdvance(val);
    return this._tok('TEXT', val, start);
  }

  // ---- helpers ----
  private _slice(start: number, end: number): string {
    return this.input.slice(start, Math.min(end, this.input.length));
  }

  private _advance(n: number, isNewline = false) {
    this.i += n;
    if (isNewline) {
      this.line += 1;
      this.col = 1;
    } else {
      this.col += n;
    }
  }

  private _fastAdvance(chunk: string) {
    const parts = chunk.split(this.delims.segment);
    if (parts.length > 1) {
      this.line += parts.length - 1;
      this.col = (parts.at(-1)?.length ?? 0) + 1;
    } else {
      this.col += chunk.length;
    }
    this.i += chunk.length;
  }

  private _tok(
    type: TokenType,
    value: string | undefined,
    start: Position['start']
  ): Token {
    const end = { offset: this.i, line: this.line, column: this.col };
    return { type, value, position: { start, end } };
  }

  // Iterable protocol (sync)
  [Symbol.iterator](): Iterator<Token> {
    return {
      next: () => {
        const t = this.next();
        if (t == null) {
          return { done: true, value: undefined as unknown as Token };
        }
        return { done: false, value: t };
      },
    };
  }
}
