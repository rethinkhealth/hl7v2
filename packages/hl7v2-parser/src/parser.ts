import type { Root } from '@rethinkhealth/hl7v2-ast';
import { DEFAULT_DELIMITERS } from '@rethinkhealth/hl7v2-utils';
import type { Plugin } from 'unified';
import { defaultPreprocessors, runPreprocessors } from './preprocessor';
import { parseHL7v2FromIterator } from './processor';
import { HL7v2Tokenizer } from './tokenizer';
import type { ParseOptions, ParserContext, Token, Tokenizer } from './types';

function* iterateTokenizerSync(t: Tokenizer): Iterable<Token> {
  for (let tok = t.next(); tok; tok = t.next()) {
    yield tok;
  }
}

export function parseHL7v2(input: string, opts: ParseOptions): Root {
  let ctx: ParserContext = {
    input,
    delimiters: opts.delimiters || DEFAULT_DELIMITERS,
  };
  // Run preprocessing
  ctx = runPreprocessors(ctx, opts.preprocess || defaultPreprocessors);
  // Run tokenizer
  const tokenizer = new HL7v2Tokenizer();
  // Reset tokenizer.
  tokenizer.reset(ctx);
  // Parse
  return parseHL7v2FromIterator(iterateTokenizerSync(tokenizer), ctx);
}

const hl7v2Parser: Plugin<[ParseOptions?], string, Root> = function (
  options = {}
) {
  this.parser = (document: string) => parseHL7v2(document, options);
};

export default hl7v2Parser;
