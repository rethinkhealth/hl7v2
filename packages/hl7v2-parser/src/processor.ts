import type { HL7v2Node } from '@rethinkhealth/hl7v2-ast';
import type { Plugin } from 'unified';
import { fromHL7v2Pipeline } from './parser';
import type { ParseOptions } from './types';

const hl7v2Parser: Plugin<[ParseOptions?], undefined, HL7v2Node> = function (
  options: ParseOptions = {}
): void {
  // biome-ignore lint/complexity/noUselessThisAlias: this is a plugin
  const self = this;

  function parser(this: unknown, value: string): HL7v2Node {
    return fromHL7v2Pipeline(value, options);
  }

  self.parser = parser;
};

export default hl7v2Parser;
