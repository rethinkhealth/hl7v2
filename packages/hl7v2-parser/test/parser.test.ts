import type {
  Component,
  Field,
  FieldRepetition,
  Root,
  Segment,
  Subcomponent,
} from '@rethinkhealth/hl7v2-ast';
import type { HL7v2Delimiters } from '@rethinkhealth/hl7v2-utils';
import { describe, expect, it } from 'vitest';
import { parseHL7v2 } from '../src/parser';

const delims: HL7v2Delimiters = {
  field: '|',
  component: '^',
  repetition: '~',
  escape: '\\',
  subcomponent: '&',
  segment: '\r',
};

const asSeg = (n: Root['children'][number]): Segment => n as Segment;
const asField = (n: Segment['children'][number]): Field => n as Field;
const asRep = (n: Field['children'][number]): FieldRepetition =>
  n as FieldRepetition;
const asComp = (n: FieldRepetition['children'][number]): Component =>
  n as Component;
const asSub = (n: Component['children'][number]): Subcomponent =>
  n as Subcomponent;

describe('parser (string input via HL7v2Tokenizer)', () => {
  it('parses a single segment with two fields', () => {
    const root = parseHL7v2('PID|1\r', { delimiters: delims });
    expect(root.children).toHaveLength(1);
    const seg = asSeg(root.children[0]);
    expect(seg.children).toHaveLength(2);
    expect(
      asSub(
        asComp(asRep(asField(seg.children[0]).children[0]).children[0])
          .children[0]
      ).value
    ).toBe('PID');
    expect(
      asSub(
        asComp(asRep(asField(seg.children[1]).children[0]).children[0])
          .children[0]
      ).value
    ).toBe('1');

    expect(root).toMatchSnapshot();
  });

  it('parses a segment ending with a field delimiter', () => {
    const root = parseHL7v2('PID|1|', { delimiters: delims });
    expect(root.children).toHaveLength(1);
    const seg = asSeg(root.children[0]);
    expect(seg.children).toHaveLength(2);
    expect(
      asSub(
        asComp(asRep(asField(seg.children[0]).children[0]).children[0])
          .children[0]
      ).value
    ).toBe('PID');
    expect(
      asSub(
        asComp(asRep(asField(seg.children[1]).children[0]).children[0])
          .children[0]
      ).value
    ).toBe('1');

    expect(root).toMatchSnapshot();
  });

  it('parses a segment ending with multiple field delimiters', () => {
    const root = parseHL7v2('PID|1|||||', { delimiters: delims });
    expect(root.children).toHaveLength(1);
    const seg = asSeg(root.children[0]);
    expect(seg.children).toHaveLength(6);

    expect(root).toMatchSnapshot();
  });

  it('parses a segment ending with multiple field delimiters and segment carriage return', () => {
    const root = parseHL7v2('PID|1|||||\r', { delimiters: delims });
    const root2 = parseHL7v2('PID|1|||||', { delimiters: delims });
    expect(root.children).toHaveLength(1);
    const seg = asSeg(root.children[0]);
    expect(seg.children).toHaveLength(6);

    expect(root).toEqual(root2);
    expect(root).toMatchSnapshot();
  });

  it('parses a segment ending with two field delimiters and preserves the last empty field', () => {
    const root = parseHL7v2('PID|1||', { delimiters: delims });
    const seg = asSeg(root.children[0]);

    // Expect 1 segment at the root
    expect(root.children).toHaveLength(1);

    // Expect 3 field in the segment
    expect(seg.children).toHaveLength(3);

    // Expect the last field to be empty array of field repetitions
    const lastField = asField(seg.children[2]);
    expect(lastField.children).toEqual([
      {
        type: 'field-repetition',
        children: [
          {
            type: 'component',
            children: [],
          },
        ],
      },
    ]);

    // Root should match the snapshot as a general check
    expect(root).toMatchSnapshot();
  });

  it('equates with and without CR when ending with a single trailing field delimiter', () => {
    const withCr = parseHL7v2('PID|1|\r', { delimiters: delims });
    const withoutCr = parseHL7v2('PID|1|', { delimiters: delims });
    expect(withCr).toEqual(withoutCr);
    const seg = asSeg(withCr.children[0]);
    expect(seg.children).toHaveLength(2);
  });

  it('handles leading double field delimiters as two empty leading fields', () => {
    const root = parseHL7v2('||A\r', { delimiters: delims });
    const seg = asSeg(root.children[0]);
    expect(seg.children).toHaveLength(3);

    expect(root).toMatchSnapshot();
  });

  it('preserves trailing component delimiter by creating an empty component', () => {
    const root = parseHL7v2('PID|A^\r', { delimiters: delims });
    const seg = asSeg(root.children[0]);
    const field2 = asField(seg.children[1]);
    const rep = asRep(field2.children[0]);
    expect(rep.children.length).toBe(2);
    // Second component exists but has no subcomponents
    expect(asComp(rep.children[1]).children.length).toBe(0);

    expect(root).toMatchSnapshot();
  });

  it('preserves trailing subcomponent delimiter by creating an empty subcomponent', () => {
    const root = parseHL7v2('PID|A&B&\r', { delimiters: delims });
    const seg = asSeg(root.children[0]);
    const field2 = asField(seg.children[1]);
    const rep = asRep(field2.children[0]);
    const comp = asComp(rep.children[0]);
    expect(comp.children.map((s) => s.value)).toEqual(['A', 'B', '']);
  });

  it('preserves trailing repetition delimiter by creating an empty repetition container', () => {
    const root = parseHL7v2('OBX|1~2~\r', { delimiters: delims });
    const seg = asSeg(root.children[0]);
    const field2 = asField(seg.children[1]);
    expect(field2.children.length).toBe(3);
    // Last repetition has a component container but no subcomponents
    const lastRep = asRep(field2.children.at(-1) as FieldRepetition);
    expect(asComp(lastRep.children[0]).children.length).toBe(0);
  });

  it('parses MSH line with bootstrap: TEXT("MSH"), FIELD_DELIM, TEXT(MSH.2)', () => {
    const root = parseHL7v2('MSH|^~\\&|SENDER\r', { delimiters: delims });
    const seg = asSeg(root.children[0]);
    expect(seg.children).toHaveLength(3);
    expect(
      asSub(
        asComp(asRep(asField(seg.children[0]).children[0]).children[0])
          .children[0]
      ).value
    ).toBe('MSH');
    expect(
      asSub(
        asComp(asRep(asField(seg.children[1]).children[0]).children[0])
          .children[0]
      ).value
    ).toBe('^~\\&');
    expect(
      asSub(
        asComp(asRep(asField(seg.children[2]).children[0]).children[0])
          .children[0]
      ).value
    ).toBe('SENDER');
  });

  it('handles components and subcomponents', () => {
    const root = parseHL7v2('PID|1^A&B\r', { delimiters: delims });
    const seg = asSeg(root.children[0]);
    const field2 = asField(seg.children[1]);
    const rep = asRep(field2.children[0]);
    expect(rep.children).toHaveLength(2);
    expect(asSub(rep.children[0].children[0]).value).toBe('1');
    const comp2 = asComp(rep.children[1]);
    expect(comp2.children.map((s) => s.value)).toEqual(['A', 'B']);
  });

  it('handles repetitions', () => {
    const root = parseHL7v2('OBX|1~2\r', { delimiters: delims });
    const seg = asSeg(root.children[0]);
    const field2 = asField(seg.children[1]);
    expect(field2.children).toHaveLength(2);
    expect(
      asSub(asComp(asRep(field2.children[0]).children[0]).children[0]).value
    ).toBe('1');
    expect(
      asSub(asComp(asRep(field2.children[1]).children[0]).children[0]).value
    ).toBe('2');
  });

  it('handles leading field delimiter as empty first field', () => {
    const root = parseHL7v2('|A\r', { delimiters: delims });
    const seg = asSeg(root.children[0]);
    expect(seg.children).toHaveLength(2);
    expect(
      asSub(
        asComp(asRep(asField(seg.children[0]).children[0]).children[0])
          .children[0]
      ).value
    ).toBe('');
    expect(
      asSub(
        asComp(asRep(asField(seg.children[1]).children[0]).children[0])
          .children[0]
      ).value
    ).toBe('A');
  });

  it('finalizes last segment without trailing segment delimiter', () => {
    const root = parseHL7v2('PID|1', { delimiters: delims });
    expect(root.children).toHaveLength(1);
    const seg = asSeg(root.children[0]);
    expect(seg.children).toHaveLength(2);
  });
});
