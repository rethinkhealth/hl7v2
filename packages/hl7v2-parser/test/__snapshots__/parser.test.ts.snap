// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parser (string input via HL7v2Tokenizer) > handles leading double field delimiters as two empty leading fields 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "",
                    },
                  ],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "A",
                    },
                  ],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
      ],
      "type": "segment",
    },
  ],
  "type": "root",
}
`;

exports[`parser (string input via HL7v2Tokenizer) > parses a segment ending with a field delimiter 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "PID",
                    },
                  ],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "1",
                    },
                  ],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
      ],
      "type": "segment",
    },
  ],
  "type": "root",
}
`;

exports[`parser (string input via HL7v2Tokenizer) > parses a segment ending with multiple field delimiters 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "PID",
                    },
                  ],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "1",
                    },
                  ],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
      ],
      "type": "segment",
    },
  ],
  "type": "root",
}
`;

exports[`parser (string input via HL7v2Tokenizer) > parses a segment ending with multiple field delimiters and segment carriage return 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "PID",
                    },
                  ],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "1",
                    },
                  ],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
      ],
      "type": "segment",
    },
  ],
  "type": "root",
}
`;

exports[`parser (string input via HL7v2Tokenizer) > parses a segment ending with two field delimiters and preserves the last empty field 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "PID",
                    },
                  ],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "1",
                    },
                  ],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
      ],
      "type": "segment",
    },
  ],
  "type": "root",
}
`;

exports[`parser (string input via HL7v2Tokenizer) > parses a single segment with two fields 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "PID",
                    },
                  ],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "1",
                    },
                  ],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
      ],
      "type": "segment",
    },
  ],
  "type": "root",
}
`;

exports[`parser (string input via HL7v2Tokenizer) > preserves trailing component delimiter by creating an empty component 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "PID",
                    },
                  ],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "A",
                    },
                  ],
                  "type": "component",
                },
                {
                  "children": [],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
      ],
      "type": "segment",
    },
  ],
  "type": "root",
}
`;
